text/0069-api-v2.md

```diff
@@ -0,0 +1,242 @@
+# TiKV API V2
+
+## Motivation
+
+`API V2` is a set of breaking changes that aim to solve serval issues with current RawKV (hereafter referred to as `API V1`):
+
+1. RawKV is not safe to use along with TxnKV. By solving this, TiDB will be able to support RawKV as the table's storage engine, which will enrich TiDB's use cases.
+2. RawKV TTL is controlled by Store configuration. Switching the configuration will cause data corruption in silence.
+3. RawKV TTL is encoded into the value by appending 8-bytes UNIX timestamp to the end of the value, therefore it's hard to introduce other encode afterward.
+4. RawKV compare_and_swap is easy to make mistake.
+5. It could be nice if we can deploy multiple applications on one TiKV cluster.
+
+## Detailed Design
+
+### New key-value codec
+
+This RFC introduces a new key encode to RawKV and TxnKV, and a new value encode to RawKV, which will allow the RawKV to be used along with TxnKV and also allow TiKV to flexibly add meta, e.g. the TTL, to a RawKV value.
+
+In addition, keys will be contained in keyspaces, where the keys in different keyspace are totally independant. If keyspace is not specified, the keyspace 'default' will be used.
+
+The `API V2` is enabled by a switch on PD. Since it changes the storage encode, it will be not compatible to switch between `API V1` and `API V2` while there are non-TiDB data in TiKV. TiDB data is specially treated in order to not be affected by the change.
+
+#### Key Encode
+
+Once the `API V2` is enabled, the key will be starting with either:
+
+1. `m` or `t`: TxnKV key. Used by TiDB.
+2. `k{keyspace prefix id}x`: TxnKV key.
+3. `k{keyspace prefix id}r`: RawKV key.
+4. `k{keyspace prefix id}a`: RawKV atomic key, on which you can use `CompareAndSwap`.
```

Comment:
Why need a standalone keyspace?

Reply:
It's not a standalone keyspace. it's an individual mode to plain rawkv mode in a keyspace. This means the client would provide three kinds of clients to access the same keyspace (RawClient, RawAtomicClient, TxnClient), and data for each client is transparent to other clients.

Reply:
Why need to separate the data physically? In theory, RawKV should be able to accessed atomically without changing the data.

Reply:
+1

Reply:
The case of corruption is when both transaction and RawKV perform writes? So we can allow read-only RawKV access? We can state these access conditions in the metadata for the keyspace and we could actually enforce these access conditions in the clients now and to some extent enforce them in the servers, I am not familiar enough to say how much.

Without having some kind of enforcement there will be concerns about corrupting data.

Reply:
The case is the limitation of the current implementation of CAS. 

Currently, TiKV archives CAS on a key by believing the user would tell TiKV whether the keys in `RawPut` and `RawDelete` requests may be `RawCAS` in order to acquire a latch during the processing. If the user gives wrong information that a key will not be CAS when `RawPut`, and sends `RawCAS` later, the `RawCAS` is not guaranteed.

There are three solutions to this problem:

1. Force `RawPut` and `RawDelete` on every key to acquiring a latch. This will significantly reduce the throughput when CAS is not used.

2. Force `RawPut` and `RawDelete` on `k{keyspace id}a` to acquiring a latch. (In this RFC)

3. Implement `RawCAS` in Raft apply. I'm still drafting an RFC for it.